### PR TITLE
fix concurrency problem + minor refactoring

### PR DESCRIPTION
**Pull request by Aleksandr-Filichkin**

### CODE REVIEW

**Comment by Aleksandr-Filichkin** [comment]
@albertattard for review

File: src/main/java/uk/tw/energy/service/PricePlanService.java [hunk 1]
```diff
@@ -52,14 +52,18 @@ private BigDecimal calculateAverageReading(List<ElectricityReading> electricityR
...
+            if (startTime.isAfter(electricityReading.getTime())) {
+                startTime = electricityReading.getTime();
+            }
+            if (endTime.isBefore(electricityReading.getTime())) {
+                endTime = electricityReading.getTime();
+            }
+        }
 
-        return BigDecimal.valueOf(Duration.between(first.getTime(), last.getTime()).getSeconds() / 3600.0);
+        return BigDecimal.valueOf(Duration.between(startTime, endTime).getSeconds() / 3600.0);
```
Comment (xpmatteo) [inline src/main/java/uk/tw/energy/service/PricePlanService.java:66]: The original version seems more declarative and clear to me.

File: src/main/java/uk/tw/energy/service/MeterReadingService.java [hunk 2]
```diff
@@ -22,9 +22,12 @@ public MeterReadingService(Map<String, List<ElectricityReading>> meterAssociated
...
-            meterAssociatedReadings.put(smartMeterId, new ArrayList<>());
-        }
-        meterAssociatedReadings.get(smartMeterId).addAll(electricityReadings);
+        meterAssociatedReadings.compute(smartMeterId, (key, value) -> {
+            if (value == null) {
+                value = new ArrayList<>();
+            }
+            value.addAll(electricityReadings);
+            return value;
+        });
```
Comment (xpmatteo) [inline src/main/java/uk/tw/energy/service/MeterReadingService.java:31]: I think this version is *much* harder to read than the original. The original is obvious, this one is clever, and clever is bad.  The candidate is already under pressure when we do the interview!!!

**Comment by albertattard** [comment]
Thank you for your contribution.  Unfortunately we are limiting the changes to the codebase, as changes make it harder for interviewers.

**Pull request closed without merging**